pkg/rclone/aws.go

```diff
@@ -5,42 +5,82 @@ package rclone
 import (
 	"context"
 	"encoding/json"
+	"io"
 	"net/http"
 	"time"
 
+	"github.com/pkg/errors"
 	"github.com/rclone/rclone/fs"
 )
 
-// awsRegionFromMetadataAPI uses instance metadata API to fetch region of the
-// running instance see
-// https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html
+// awsRegionFromMetadataAPI uses instance metadata API v2 to fetch region of the
+// running instance see https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-retrieval.html
 // Returns empty string if region can't be obtained for whatever reason.
+// Fallbacks to IMDSv1 if the session token cannot be obtained.
 func awsRegionFromMetadataAPI() string {
-	const url = "http://169.254.169.254/latest/dynamic/instance-identity/document"
+	const (
+		tokenURL = "http://169.254.169.254/latest/api/token"
+		docURL   = "http://169.254.169.254/latest/dynamic/instance-identity/document"
+	)
 
-	req, err := http.NewRequestWithContext(context.Background(), http.MethodGet, url, http.NoBody)
+	// Step 1: Request an IMDSv2 session token
+	token, err := func() (string, error) {
+		reqToken, err := http.NewRequestWithContext(context.Background(), http.MethodPut, tokenURL, http.NoBody)
+		if err != nil {
+			return "", errors.Wrap(err, "create token request")
+		}
+		reqToken.Header.Set("X-aws-ec2-metadata-token-ttl-seconds", "21600")
+		tokenClient := http.Client{
+			Timeout: 2 * time.Second,
+		}
+		resToken, err := tokenClient.Do(reqToken)
+		if err != nil {
+			return "", errors.Wrap(err, "IMDSv2 failed to fetch session token")
+		}
+		defer resToken.Body.Close()
+
+		if resToken.StatusCode != http.StatusOK {
+			return "", errors.Wrap(err, "failed to retrieve session token")
+		}
+		token, err := io.ReadAll(resToken.Body)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to read session token")
+		}
+		return string(token), nil
+	}()
```

Comment:
nit: could we extract it to a separate function?

Reply:
we could, done